File: hibernate-scala-examples/src/main/scala/org/hibernate/examples/hibernate/interceptor/MultipleInterceptor.scala
package org.hibernate.examples.interceptor

import java.io.Serializable
import java.util
import org.hibernate.`type`.Type
import org.hibernate.{Interceptor, EmptyInterceptor}
import scala.collection.mutable.ArrayBuffer

/**
 * kr.debop4s.data.hibernate.interceptor.MultipleInterceptor
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 12. 24. 오후 4:43
 */
class MultipleInterceptor extends EmptyInterceptor {

    val interceptors = ArrayBuffer[Interceptor]()

    def this(interceptors: Interceptor*) {
        this()
        this.interceptors ++= interceptors
    }

    def addInterceptor(interceptor: Interceptor) {
        this.interceptors += interceptor
    }

    def removeInterceptor(interceptor: Interceptor) {
        this.interceptors -= interceptor
    }

    def exists: Boolean = interceptors != null && interceptors.size > 0

    override def onDelete(entity: Any, id: Serializable, state: Array[AnyRef], propertyNames: Array[String], types: Array[Type]) {
        if (exists) {
            interceptors.foreach(x => x.onDelete(entity, id, state, propertyNames, types))
        }
    }

    override def onFlushDirty(entity: Any, id: Serializable, currentState: Array[AnyRef], previousState: Array[AnyRef], propertyNames: Array[String], types: Array[Type]): Boolean = {
        if (exists) {
            interceptors.foreach(x => x.onFlushDirty(entity, id, currentState, previousState, propertyNames, types))
        }
        false
    }

    override def onSave(entity: scala.Any, id: Serializable, state: Array[AnyRef], propertyNames: Array[String], types: Array[Type]): Boolean = {
        if (exists) {
            interceptors.foreach(x => x.onSave(entity, id, state, propertyNames, types))
        }
        false
    }

    override def onLoad(entity: scala.Any, id: Serializable, state: Array[AnyRef], propertyNames: Array[String], types: Array[Type]): Boolean = {
        if (exists) {
            interceptors.foreach(x => x.onLoad(entity, id, state, propertyNames, types))
        }
        false
    }

    override def postFlush(entities: util.Iterator[_]) {
        if (exists) {
            interceptors.foreach(x => x.postFlush(entities))
        }
    }

    override def preFlush(entities: util.Iterator[_]) {
        if (exists) {
            interceptors.foreach(x => x.preFlush(entities))
        }
    }
}

File: hibernate-examples/src/main/java/org/hibernate/examples/hibernate/interceptor/MultipleInterceptor.java
package org.hibernate.examples.hibernate.interceptor;

import lombok.extern.slf4j.Slf4j;
import org.hibernate.EmptyInterceptor;
import org.hibernate.Interceptor;
import org.hibernate.type.Type;

import java.io.Serializable;
import java.util.*;

/**
 * Hibernate에 등록할 수 있는 Interceptor는 하나입니다.
 * MultipleInterceptor는 복수의 Interceptor이 작업할 수 있도록 해줍니다.
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 2013. 11. 28. 오후 1:51
 */
@Slf4j
public class MultipleInterceptor extends EmptyInterceptor {

  private final List<Interceptor> interceptors = new ArrayList<Interceptor>();

  public MultipleInterceptor(Collection<? extends Interceptor> c) {
    this.interceptors.addAll(c);
  }

  public MultipleInterceptor(Interceptor... array) {
    Collections.addAll(this.interceptors, array);
  }

  public void addInterceptor(Interceptor interceptor) {
    this.interceptors.add(interceptor);
  }

  public void removeInterceptor(Interceptor interceptor) {
    this.interceptors.remove(interceptor);
  }

  public boolean exists() {
    return interceptors.size() > 0;
  }

  @Override
  public void onDelete(Object entity, Serializable id, Object[] state, String[] propertyNames, Type[] types) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.onDelete(entity, id, state, propertyNames, types);
      }
    }
  }

  @Override
  public boolean onFlushDirty(Object entity, Serializable id, Object[] currentState, Object[] previousState, String[] propertyNames, Type[] types) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.onFlushDirty(entity, id, currentState, previousState, propertyNames, types);
      }
    }
    return false;
  }

  @Override
  public boolean onSave(Object entity, Serializable id, Object[] state, String[] propertyNames, Type[] types) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.onSave(entity, id, state, propertyNames, types);
      }
    }
    return false;
  }

  @Override
  public boolean onLoad(Object entity, Serializable id, Object[] state, String[] propertyNames, Type[] types) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.onLoad(entity, id, state, propertyNames, types);
      }
    }
    return false;
  }

  @Override
  public void postFlush(Iterator entities) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.postFlush(entities);
      }
    }
  }

  @Override
  public void preFlush(Iterator entities) {
    if (exists()) {
      for (Interceptor interceptor : interceptors) {
        interceptor.preFlush(entities);
      }
    }
  }

  private static final long serialVersionUID = -818808374889268894L;
}
